graphein/datasets/pdb_data.py

```diff
@@ -0,0 +1,1092 @@
+        :type overwrite: bool, optional
+        :param max_workers: Number of processes to use, defaults to ``8``.
+        :type max_workers: int, optional
+        :param chunksize: Chunk size for each worker, defaults to ``32``.
+        :type chunksize: int, optional
+        """
+        log.info(f"Downloading {len(self.unique_pdbs)} PDB files...")
+        download_pdb_multiprocessing(
+            self.unique_pdbs,
+            out_dir,
+            overwrite=overwrite,
+            max_workers=max_workers,
+            chunksize=chunksize,
+        )
+
+    def write_chains(self) -> List[Path]:
+        """Write chains in current selection to disk. e.g., we create a file
+        of the form ``4hbb_A.pdb`` for chain ``A`` of PDB file ``4hhb.pdb``.
+
+        If the PDB files are not contained in ``self.pdb_dir``, they are
+        downloaded.
+
+        :return: List of paths to written files.
+        :rtype: List[Path]
+        """
+        # Get dictionary of PDB code : List[Chains]
+        df = self.df.groupby("pdb")["chain"].agg(list).to_dict()
+
+        # Check we have all source PDB files
+        downloaded = os.listdir(self.pdb_dir)
+        downloaded = [f for f in downloaded if f.endswith(".pdb")]
+
+        if to_download := [
+            k for k in df.keys() if f"{k}.pdb" not in downloaded
+        ]:
+            log.info(f"Downloading {len(to_download)} PDB files...")
+            download_pdb_multiprocessing(
+                to_download, self.pdb_dir, overwrite=True
+            )
+            log.info("Done downloading PDB files")
+
+        # Iterate over dictionary and write chains to separate files
+        log.info("Extracting chains...")
+        paths = []
+        for k, v in tqdm(df.items()):
+            in_file = os.path.join(self.pdb_dir, f"{k}.pdb")
+            paths.append(
+                extract_chains_to_file(in_file, v, out_dir=self.pdb_dir)
+            )
+        log.info("Done extracting chains")
+
+        # Flatten list of paths
+        return [Path(num) for sublist in paths for num in sublist]
+
+    def reset(self) -> pd.DataFrame:
+        """Reset the dataset to the original DataFrame source.
+
+        :return: The source dataset DataFrame.
+        :rtype: pd.DataFrame
+        """
+        self.df = self.source.copy()
+        return self.df
+
+    def split_df_proportionally(
+        self,
+        df: pd.DataFrame,
+        splits: List[str],
+        split_ratios: List[float],
+        assign_leftover_rows_to_split_n: int = 0,
+        random_state: int = 42,
+    ) -> Dict[str, pd.DataFrame]:
+        """Split the provided DataFrame iteratively according to given proportions.
+
+        :param df: DataFrame to split.
+        :type df: pd.DataFrame
+        :param splits: Names of splits into which to divide the provided DataFrame.
+        :type splits: List[str]
+        :param split_ratios: Ratios by which to split the provided DataFrame.
```

Comment:
Some input sanitation would be good.

Assert len(splits) == len(split_ratios)

Asset sum(split_ratios) == 1

Reply:
Addressed in [this commit](https://github.com/a-r-j/graphein/pull/272/commits/3b42dab1be3429af2e9b2c5da59166bc722c8046).